File: src/main/java/com/atsuishio/superbwarfare/client/renderer/animations/SenpaiStates.java
package com.atsuishio.superbwarfare.client.renderer.animations;

import com.github.mcmodderanchor.simplebedrockmodel.v1.common.animation.SimpleAnimationState;
import com.github.mcmodderanchor.simplebedrockmodel.v1.common.animation.SimpleTransition;

public class SenpaiStates {
    public static final SimpleAnimationState<SenpaiContext> IDLE = new SimpleAnimationState.Builder<SenpaiContext>()
            .onEnter((ctx, state) -> ctx.playAnimation("animation.senpai.idle", AnimationPlayType.LOOP))
            .evaluatePose(SenpaiContext::getPose)
            .build();

    public static final SimpleAnimationState<SenpaiContext> WALK = new SimpleAnimationState.Builder<SenpaiContext>()
            .onEnter((ctx, state) -> ctx.playAnimation("animation.senpai.walk", AnimationPlayType.LOOP))
            .evaluatePose(SenpaiContext::getPose)
            .build();

    public static final SimpleAnimationState<SenpaiContext> RUN = new SimpleAnimationState.Builder<SenpaiContext>()
            .onEnter((ctx, state) -> {
                if (ctx.isRunner()) {
                    ctx.playAnimation("animation.senpai.run2", AnimationPlayType.LOOP);
                } else {
                    ctx.playAnimation("animation.senpai.run", AnimationPlayType.LOOP);
                }
            })
            .evaluatePose(SenpaiContext::getPose)
            .build();

    public static final SimpleAnimationState<SenpaiContext> DIE = new SimpleAnimationState.Builder<SenpaiContext>()
            .onEnter((ctx, state) -> ctx.playAnimation("animation.senpai.die", AnimationPlayType.PLAY_ONCE_HOLD))
            .evaluatePose(SenpaiContext::getPose)
            .build();

    public static final SimpleTransition<SenpaiContext> TO_IDLE = new SimpleTransition.Builder<SenpaiContext>()
            .predicate(ctx -> !ctx.isMoving())
            .target(IDLE)
            .from(WALK, RUN)
            .duration(0.6f)
            .build();

    public static final SimpleTransition<SenpaiContext> TO_WALK = new SimpleTransition.Builder<SenpaiContext>()
            .predicate(ctx -> {
                var entity = ctx.getEntity();
                float limbSwingAmount = entity.walkAnimation.speed(ctx.partialTick);
                return (ctx.isMoving() || !(limbSwingAmount > -0.15f && limbSwingAmount < 0.15f)) && !entity.isAggressive();
            })
            .target(WALK)
            .from(RUN, IDLE)
            .build();

    public static final SimpleTransition<SenpaiContext> TO_RUN = new SimpleTransition.Builder<SenpaiContext>()
            .predicate(ctx -> {
                var entity = ctx.getEntity();
                return entity.isAggressive() && ctx.isMoving();
            })
            .target(RUN)
            .from(WALK, IDLE)
            .build();

    public static final SimpleTransition<SenpaiContext> TO_DIE = new SimpleTransition.Builder<SenpaiContext>()
            .predicate(ctx -> ctx.getEntity().isDeadOrDying())
            .target(DIE)
            .from(RUN, WALK, IDLE)
            .build();
}

File: src/main/kotlin/com/atsuishio/superbwarfare/client/animation/entity/SenpaiStates.kt
package com.atsuishio.superbwarfare.client.animation.entity

import com.atsuishio.superbwarfare.client.animation.AnimationPlayType
import com.github.mcmodderanchor.simplebedrockmodel.v1.common.animation.SimpleAnimationState
import com.github.mcmodderanchor.simplebedrockmodel.v1.common.animation.SimpleTransition

object SenpaiStates {
    val IDLE: SimpleAnimationState<SenpaiContext> = SimpleAnimationState.Builder<SenpaiContext>()
        .onEnter { ctx, _ -> ctx.playAnimation("animation.senpai.idle", AnimationPlayType.LOOP) }
        .evaluatePose { it.getPose() }
        .build()

    val WALK: SimpleAnimationState<SenpaiContext> = SimpleAnimationState.Builder<SenpaiContext>()
        .onEnter { ctx, _ -> ctx.playAnimation("animation.senpai.walk", AnimationPlayType.LOOP) }
        .evaluatePose { it.getPose() }
        .build()

    val RUN: SimpleAnimationState<SenpaiContext> = SimpleAnimationState.Builder<SenpaiContext>()
        .onEnter { ctx, _ ->
            if (ctx.isRunner()) {
                ctx.playAnimation("animation.senpai.run2", AnimationPlayType.LOOP)
            } else {
                ctx.playAnimation("animation.senpai.run", AnimationPlayType.LOOP)
            }
        }
        .evaluatePose { it.getPose() }
        .build()

    val DIE: SimpleAnimationState<SenpaiContext> = SimpleAnimationState.Builder<SenpaiContext>()
        .onEnter { ctx, _ -> ctx.playAnimation("animation.senpai.die", AnimationPlayType.PLAY_ONCE_HOLD) }
        .evaluatePose { it.getPose() }
        .build()

    val TO_IDLE: SimpleTransition<SenpaiContext> = SimpleTransition.Builder<SenpaiContext>()
        .predicate { !it.isMoving() }
        .target(IDLE)
        .from(WALK, RUN)
        .duration(0.6f)
        .build()

    val TO_WALK: SimpleTransition<SenpaiContext> = SimpleTransition.Builder<SenpaiContext>()
        .predicate {
            val entity = it.entity
            val limbSwingAmount = entity.walkAnimation.speed(it.partialTick)
            (it.isMoving() || !(limbSwingAmount > -0.15f && limbSwingAmount < 0.15f)) && !entity.isAggressive
        }
        .target(WALK)
        .from(RUN, IDLE)
        .build()

    val TO_RUN: SimpleTransition<SenpaiContext> = SimpleTransition.Builder<SenpaiContext>()
        .predicate {
            val entity = it.entity
            entity.isAggressive && it.isMoving()
        }
        .target(RUN)
        .from(WALK, IDLE)
        .build()

    val TO_DIE: SimpleTransition<SenpaiContext> = SimpleTransition.Builder<SenpaiContext>()
        .predicate { it.entity.isDeadOrDying }
        .target(DIE)
        .from(RUN, WALK, IDLE)
        .build()
}
